From ae33cd25bd3374fca9dc501ab447646f0c3667b9 Mon Sep 17 00:00:00 2001
From: Jiang Lu <lu.jiang@windriver.com>
Date: Sat, 24 Nov 2018 19:08:51 +0800
Subject: [PATCH 7/9] arm64:invoke device shutdown routine in
 machine_crash_shutdown()

Some device need invoke a shutdown routine during kdump, add
device shutdown routine in machine_crash_shutdown() for arm64.

Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
---
 arch/arm64/kernel/machine_kexec.c |   15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/kernel/machine_kexec.c b/arch/arm64/kernel/machine_kexec.c
index 481f54a..a14a3fe 100644
--- a/arch/arm64/kernel/machine_kexec.c
+++ b/arch/arm64/kernel/machine_kexec.c
@@ -247,6 +247,11 @@ static void machine_kexec_mask_interrupts(void)
 /**
  * machine_crash_shutdown - shutdown non-crashing cpus and save registers
  */
+
+extern void qman_crash_shutdown(void);
+extern void bman_crash_shutdown(void);
+extern void fm_crash_shutdown_all(void);
+
 void machine_crash_shutdown(struct pt_regs *regs)
 {
 	local_irq_disable();
@@ -257,7 +262,15 @@ void machine_crash_shutdown(struct pt_regs *regs)
 	/* for crashing cpu */
 	crash_save_cpu(regs, smp_processor_id());
 	machine_kexec_mask_interrupts();
-
+#ifdef CONFIG_FSL_SDK_FMAN
+	fm_crash_shutdown_all();
+#endif
+#ifdef CONFIG_FSL_QMAN_CONFIG
+	qman_crash_shutdown();
+#endif
+#ifdef CONFIG_FSL_BMAN_CONFIG
+	bman_crash_shutdown();
+#endif
 	pr_info("Starting crashdump kernel...\n");
 }
 
-- 
1.7.9.5

